providers/base/bin/pipewire_test.py

```diff
@@ -0,0 +1,608 @@
+        error = self.setpoint - process_feedback
+        self._integral = self._integral + (error * dt)
+        derivative = (error - self._previous_error) / dt
+        self._previous_error = error
+        input_change = (self.kp * error) + \
+                       (self.ki * self._integral) + \
+                       (self.kd * derivative)
+        if self._change_limit and abs(input_change) > abs(self._change_limit):
+            sign = input_change / abs(input_change)
+            input_change = sign * self._change_limit
+        return input_change
+
+    def set_change_limit(self, limit):
+        """Ensures that input value changes are lower than limit.
+
+           Setting limit of zero disables this.
+        """
+        self._change_limit = limit
+
+
+class VolumeController(object):
+    pw_types = {'input': 'SOURCE', 'output': 'SINK'}
+
+    def __init__(self, type, logger=None):
+        """Initializes the volume controller.
+
+           Arguments:
+           type: either input or output
+
+        """
+        self.type = type
+        self._volume = None
+        self.identifier = None
+        self.logger = logger
+
+    def set_volume(self, volume):
+        if not 0 <= volume <= 100:
+            return False
+        command = [
+            "wpctl",
+            "set-volume",
+            "@DEFAULT_AUDIO_%s@" % (self.pw_types[self.type]),
+            str(int(volume)) + "%",
+        ]
+        self._wpctl_output(command)
+        self._volume = volume
+        return True
+
+    def get_volume(self):
+        return self._volume
+
+    def mute(self, mute):
+        mute = str(int(mute))
+        command = [
+            "wpctl",
+            "set-mute",
+            "@DEFAULT_AUDIO_%s@" % (self.pw_types[self.type]),
+            mute,
+        ]
+        self._wpctl_output(command)
+        return True
+
+    def _wpctl_output(self, command):
+        # This method mainly calls wpctl (hence the name). Since wpctl may
+        # return a failure if the audio layer is not yet initialized, we will
+        # try running a few times in case of failure. All our invocations of
+        # wpctl should be "idempotent" so repeating them should not have
+        # any bad effects.
+        for attempt in range(0, 3):
+            try:
+                return subprocess.check_output(command,
+                                               universal_newlines=True)
+            except (subprocess.CalledProcessError):
+                time.sleep(5)
+            except FileNotFoundError:
+                break
+        self.logger.error("Fail to execute: {}".format(command))
+        sys.exit(1)
+
+
+class FileDumper(object):
+    def write_to_file(self, filename, data):
+        try:
+            with open(filename, "w") as f:
+                for i in data:
+                    print(i, file=f)
+            return_value = True
+        except (TypeError, IOError):
+            return_value = False
+        return return_value
```

Comment:
Might as well provide the error msg here, something like:
except IOError as e:
            print("ERROR: could not find %s: %s" %
                  (filename, repr(e)))